src/System.Net.Sockets/src/System/Net/Sockets/IOControlKeepAlive.cs

```diff
@@ -0,0 +1,98 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System.Net.Sockets
+{
+    internal class IOControlKeepAlive
+    {
+        private const uint WindowsDefaultTimeMs = 7200000u;
+        private const uint WindowsDefaultIntervalMs = 1000u;
+        private static readonly bool s_supportsKeepAliveViaSocketOption = SupportsKeepAliveViaSocketOption();
+        private static readonly ConditionalWeakTable<SafeSocketHandle, IOControlKeepAlive> s_socketKeepAliveTable = new ConditionalWeakTable<SafeSocketHandle, IOControlKeepAlive>();
+        [ThreadStatic]
+        private static byte[] s_keepAliveValuesBuffer;
+
+        private uint _timeMs = WindowsDefaultTimeMs;
+        private uint _intervalMs = WindowsDefaultIntervalMs;
+
+        public static bool IsNeeded
+        {
+            get
+            {
+                return !s_supportsKeepAliveViaSocketOption;
+            }
+        }
+
+        public static int Get(SafeSocketHandle handle, SocketOptionName optionName)
+        {
+            IOControlKeepAlive ioControlKeepAlive;
+            s_socketKeepAliveTable.TryGetValue(handle, out ioControlKeepAlive);
+            return optionName == SocketOptionName.TcpKeepAliveTime ?
+                MillisecondsToSeconds(ioControlKeepAlive?._timeMs ?? WindowsDefaultTimeMs) :
+                MillisecondsToSeconds(ioControlKeepAlive?._intervalMs ?? WindowsDefaultIntervalMs);
+        }
+
+        public static SocketError Set(SafeSocketHandle handle, SocketOptionName optionName, int optionValueSeconds)
+        {
+            IOControlKeepAlive ioControlKeepAlive = s_socketKeepAliveTable.GetOrCreateValue(handle);
+            if (optionName == SocketOptionName.TcpKeepAliveTime)
+                ioControlKeepAlive._timeMs = SecondsToMilliseconds(optionValueSeconds);
+            else
+                ioControlKeepAlive._intervalMs = SecondsToMilliseconds(optionValueSeconds);
+
+            byte[] buffer = s_keepAliveValuesBuffer ?? (s_keepAliveValuesBuffer = new byte[3 * sizeof(uint)]);
+            ioControlKeepAlive.Fill(buffer);
+            int realOptionLength = 0;            
+            return SocketPal.WindowsIoctl(handle, unchecked((int)IOControlCode.KeepAliveValues), buffer, null, out realOptionLength);
+        }
+
+        private static bool SupportsKeepAliveViaSocketOption()
+        {
+            using (Socket socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
```

Comment:
Is it possible this might fail if IPv4 has been disabled?  I'm wondering if we might need to check for IPv4 and IPv6 separately?